File: backend-node/src/controllers/room.controller.js
const Room = require('../models/Room');

// Get all rooms
const getAllRooms = async (req, res) => {
  try {
    const { type, available, maxPrice, capacity } = req.query;
    
    const filters = {};
    if (type) filters.roomType = type;
    if (available !== undefined) filters.isAvailable = available === 'true';
    if (maxPrice) filters.maxPrice = parseFloat(maxPrice);
    if (capacity) filters.capacity = parseInt(capacity);

    const rooms = await Room.findAll(filters);

    res.json({
      success: true,
      data: { rooms }
    });
  } catch (error) {
    console.error('Get rooms error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to fetch rooms'
    });
  }
};

// Get room by ID
const getRoomById = async (req, res) => {
  try {
    const room = await Room.findById(req.params.id);
    
    if (!room) {
      return res.status(404).json({
        success: false,
        message: 'Room not found'
      });
    }

    res.json({
      success: true,
      data: { room }
    });
  } catch (error) {
    console.error('Get room error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to fetch room'
    });
  }
};

// Get available rooms for date range
const getAvailableRooms = async (req, res) => {
  try {
    const { checkIn, checkOut, type } = req.query;

    if (!checkIn || !checkOut) {
      return res.status(400).json({
        success: false,
        message: 'Check-in and check-out dates are required'
      });
    }

    const rooms = await Room.findAvailable(checkIn, checkOut, type);

    res.json({
      success: true,
      data: { rooms }
    });
  } catch (error) {
    console.error('Get available rooms error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to fetch available rooms'
    });
  }
};

// Get room types with pricing
const getRoomTypes = async (req, res) => {
  try {
    const roomTypes = await Room.getRoomTypes();

    res.json({
      success: true,
      data: { roomTypes }
    });
  } catch (error) {
    console.error('Get room types error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to fetch room types'
    });
  }
};

// Create room (admin only)
const createRoom = async (req, res) => {
  try {
    const roomData = req.body;
    const room = await Room.create(roomData);

    res.status(201).json({
      success: true,
      message: 'Room created successfully',
      data: { room }
    });
  } catch (error) {
    console.error('Create room error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to create room'
    });
  }
};

// Update room (admin only)
const updateRoom = async (req, res) => {
  try {
    const room = await Room.findById(req.params.id);
    
    if (!room) {
      return res.status(404).json({
        success: false,
        message: 'Room not found'
      });
    }

    const updatedRoom = await Room.update(req.params.id, req.body);

    res.json({
      success: true,
      message: 'Room updated successfully',
      data: { room: updatedRoom }
    });
  } catch (error) {
    console.error('Update room error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to update room'
    });
  }
};

module.exports = {
  getAllRooms,
  getRoomById,
  getAvailableRooms,
  getRoomTypes,
  createRoom,
  updateRoom
};

File: backend/src/main/java/com/dphotel/controller/RoomController.java
package com.dphotel.controller;

import com.dphotel.dto.ApiResponse;
import com.dphotel.dto.OtherDtos;
import com.dphotel.entity.Room;
import com.dphotel.service.RoomService;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/rooms")
public class RoomController {

    private final RoomService roomService;

    public RoomController(RoomService roomService) {
        this.roomService = roomService;
    }

    @GetMapping
    public ResponseEntity<ApiResponse<Map<String, Object>>> getAllRooms(
            @RequestParam(required = false) String type,
            @RequestParam(required = false) Boolean available,
            @RequestParam(required = false) BigDecimal maxPrice,
            @RequestParam(required = false) Integer capacity) {
        List<Room> rooms = roomService.getAllRooms(type, available, maxPrice, capacity);
        Map<String, Object> data = new HashMap<>();
        data.put("rooms", rooms);
        return ResponseEntity.ok(ApiResponse.success(data));
    }

    @GetMapping("/types")
    public ResponseEntity<ApiResponse<Map<String, Object>>> getRoomTypes() {
        List<Map<String, Object>> roomTypes = roomService.getRoomTypes();
        Map<String, Object> data = new HashMap<>();
        data.put("roomTypes", roomTypes);
        return ResponseEntity.ok(ApiResponse.success(data));
    }

    @GetMapping("/available")
    public ResponseEntity<ApiResponse<Map<String, Object>>> getAvailableRooms(
            @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate checkIn,
            @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate checkOut,
            @RequestParam(required = false) String type) {

        if (checkIn == null || checkOut == null) {
            return ResponseEntity.badRequest()
                    .body(ApiResponse.error("Check-in and check-out dates are required"));
        }

        List<Room> rooms = roomService.getAvailableRooms(checkIn, checkOut, type);
        Map<String, Object> data = new HashMap<>();
        data.put("rooms", rooms);
        return ResponseEntity.ok(ApiResponse.success(data));
    }

    @GetMapping("/{id}")
    public ResponseEntity<ApiResponse<Map<String, Object>>> getRoomById(@PathVariable String id) {
        Room room = roomService.getRoomById(id);
        Map<String, Object> data = new HashMap<>();
        data.put("room", room);
        return ResponseEntity.ok(ApiResponse.success(data));
    }

    @PostMapping
    public ResponseEntity<ApiResponse<Map<String, Object>>> createRoom(@RequestBody OtherDtos.RoomRequest request) {
        Room room = roomService.createRoom(request);
        Map<String, Object> data = new HashMap<>();
        data.put("room", room);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.success("Room created successfully", data));
    }

    @PutMapping("/{id}")
    public ResponseEntity<ApiResponse<Map<String, Object>>> updateRoom(
            @PathVariable String id,
            @RequestBody OtherDtos.RoomRequest request) {
        Room room = roomService.updateRoom(id, request);
        Map<String, Object> data = new HashMap<>();
        data.put("room", room);
        return ResponseEntity.ok(ApiResponse.success("Room updated successfully", data));
    }
}
